system/Database/BaseBuilder.php

```diff
@@ -2051,6 +2067,72 @@ public function onConstraint($set)
         return $this;
     }
 
+    /**
+     * Sets data source as a query for insert/update/upsert
+     *
+     * @param BaseBuilder|RawSql|string $query
+     */
+    public function fromQuery($query): BaseBuilder
```

Comment:
Do we really need this public method?

Reply:
I prefer to call the method directly rather than pass it in the *Batch function. I think it is easier to read the code when you do it this way. 

```php
$this->db->table('user')->insertBatch($query);

$this->db->table('user')->fromQuery($query)->insertBatch();

$this->db->table('user')->setData($data)->insertBatch();
```

I also debated weather or not I should accept queries in the $set parameter of the *Batch() methods. $set was really intended as the parameter for setData(). But it works passing query in *Batch() methods.

At some point It might be useful for other methods as well. We could always change it to public later I suppose though.

Reply:
`fromSubquery()` works a bit differently. It sets the query to QBFrom[0] . `fromQuery()` sets the query to `QBOptions['fromQuery']` and `QBKeys` to the columns.

One option might be to move the `fromQuery()` logic to `setData()`. This kind of makes sense programming wise.. the query is an alternate to a dataset. The only problem might be is that people might try and use it where it is not supported.

Reply:
The following API seems better if it is possible:

```php
$this->db->table('user')->fromSubquery($query)->insertBatch();
```

> fromSubquery() works a bit differently. It sets the query to QBFrom[0] . fromQuery() sets the query to QBOptions['fromQuery'] and QBKeys to the columns.

It is just internal implementation. We can change the implementation.

Reply:
I have added a parameter for manually setting columns now. This differentiates it even farther from fromSubQuery(). I admit it would be nice to have one method but i'm not seeing how everything could play nicely.

I really don't like the implementation of fromSubQuery() - at least as far as the examples in the documentation.

```php
$subquery = $db->table('users')->select('id, name');
$builder  = $db->newQuery()->fromSubquery($subquery, 't');
$query    = $builder->get();
// Produces: SELECT * FROM (SELECT `id`, `name` FROM users) `t`
```
I'd rather see something like:

```php
$queryTable= $db->table('users')->select('id, name');
$queryJoin= $db->table('users2')->select('id, email');

$builder  = $db->table($db->queryAsTable($queryTable, 'alias'))
                        ->select('alias.id, alias.name, alias2.email')
                        ->join($db->queryAsTable($queryJoin, 'alias2'), 'alias2.id = alias.id', 'inner');
$query    = $builder->get();

/*
SELECT alias.id, alias.name , alias2.email
FROM (SELECT id, name FROM users) alias
INNER JOIN (SELECT id, email FROM users2) alias2 ON alias2.id = alias.id
*/

// or perhaps even:
$queryAsTable= $db->table('users')->select('id, name')->queryAsTable('alias'); // returns RawSql ?
$db->table($queryAsTable)->select('alias.id, alias.name')
```

Reply:
https://github.com/codeigniter4/CodeIgniter4/blob/978059c4fac2dbf56e5019b84427de8a7fb01b80/system/Database/BaseBuilder.php#L594-L600
https://github.com/codeigniter4/CodeIgniter4/blob/978059c4fac2dbf56e5019b84427de8a7fb01b80/system/Database/BaseBuilder.php#L2082-L2089

Reply:
Is the name `setQuery()` appropriate? Would a more specific name be better?

Reply:
setQuery() seems a logical alternative to setData(). Maybe you could do setQueryAsData() but that seems a bit long.

Reply:
If I see `setData()` without any knowledge of the implementation, I would guess it will set data for the query, that is the data to save.

But if I see `setQuery()`, I would wonder what the query is. Why do I set a query to QueryBuilder? QueryBuilder is not what to build a query?

In my opinion, `setQuery()` is not as clear about what it does as `setData()`.

I'm not sure which is better.:
```php
$this->db->table('user')->setQuery($query)->insertBatch();
```
```php
$this->db->table('user')->setQueryAsData($query)->insertBatch();
```
```php
$this->db->table('user')->setDataAsQuery($query)->insertBatch();
```
```php
$this->db->table('user')->fromQuery($query)->insertBatch();
```

Reply:
>`setDataAsQuery()`

This is what `setData()` actually does. 

`fromQuery()` was ok but it might get confused with `fromSubQuery()`

In the future I'd like to have a method that is something like `queryAsTable()` This would be similar to what `fromSubQuery()` does in that it puts the query in parenthesis and adds an alias. Basically the results of this function could go anywhere a table name could go. This method could be used instead of `fromSubQuery()` by just setting `table()`. It could also go in `join()`.

`setQueryAsData()` is the only other option I see that might make sense.

One thing to keep in mind is that the API language doesn't have to necessarily follow the SQL. I mean for the most part yes it is nice but with some of these more advanced SQL statements they vary with each DBMS. I liked fromQuery() because it seemed clear that the data was coming from the query.

Reply:
Both `fromQuery()` and `setQueryAsData()` seem right to me because they describe well what will happen.

The only disadvantage of `fromQuery()` is that it has a very similar name to an already existing method, and people may be confused about why they're working differently.

For that reason, I would go with `setQueryAsData()`.

Reply:
>For that reason, I would go with setQueryAsData().

I agree.  I think this clearly presents itself as an alternative to setData().

I changed everything to this.

@kenjis how do you feel about this?

Reply:
`setQueryAsData()` is much better than `setData()` and `setQuery()`.

Reply:
Normally with *Batch you use setData(). setQueryAsData() is a fitting alternative. It is self explanatory.